prov/rxd/src/rxd_cq.c

```diff
@@ -229,391 +239,523 @@ static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
+{
+	struct rxd_x_entry *rx_entry;
+
+	dlist_foreach_container(&ep->peers[peer].buf_cq, struct rxd_x_entry,
+				rx_entry, entry) {
+		if (rx_entry->msg_id != rxd_get_next_no(ep->peers[peer].rx_msg_id))
+			return;
+
+		rxd_complete_rx(ep, rx_entry);
+	}
+}
+
+static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
+			     struct rxd_data_pkt *pkt, size_t size)
 {
-	if (!(rx_entry->flags & RXD_INLINE) &&
-	    rx_entry->rx_id != pkt_hdr.rx_id)
-		return -FI_EALREADY;
+	struct rxd_domain *rxd_domain = rxd_ep_domain(ep);
+	uint64_t done;
+
+	done = ofi_copy_to_iov(rx_entry->iov, rx_entry->iov_count,
+			       rxd_domain->max_inline_sz +
+			       (pkt->pkt_hdr.seg_no * rxd_domain->max_seg_sz),
+			       pkt->msg, size - RXD_DATA_HDR_SIZE - ep->prefix_size);
+
+	rx_entry->bytes_done += done;
+	ep->peers[pkt->pkt_hdr.peer].rx_seq_no++;
+	rx_entry->next_seg_no++;
 
-	if ((rx_entry->tx_id == pkt_hdr.tx_id) &&
-	    (rx_entry->key == pkt_hdr.key))
+	if (rx_entry->next_seg_no < rx_entry->num_segs) {
+		if (!(ep->peers[pkt->pkt_hdr.peer].rx_seq_no %
+		    ep->peers[pkt->pkt_hdr.peer].rx_window))
+			rxd_ep_post_ack(ep, pkt->pkt_hdr.peer);
+		return;
+	}
+	rxd_ep_post_ack(ep, pkt->pkt_hdr.peer);
+
+	if (rxd_env.ooo_rdm && rx_entry->msg_id !=
+	    rxd_get_next_no(ep->peers[pkt->pkt_hdr.peer].rx_msg_id)) {
+		dlist_insert_order(&ep->peers[pkt->pkt_hdr.peer].buf_cq,
+				   &rxd_cq_order, &rx_entry->entry);
+		return;
+	}
+
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
+
+	rxd_complete_rx(ep, rx_entry);
+	if (rxd_env.ooo_rdm)
+		rxd_progress_buf_cq(ep, pkt->pkt_hdr.peer);
+
+	fastlock_release(&ep->util_ep.rx_cq->cq_lock);
+
+	rxd_ep_post_ack(ep, pkt->pkt_hdr.peer);
+}
+
+static int rxd_check_pkt_ids(struct rxd_x_entry *rx_entry,
+			     struct rxd_pkt_hdr *pkt_hdr)
+{
+	if ((rx_entry->tx_id == pkt_hdr->tx_id) &&
+	    (rx_entry->rx_id == pkt_hdr->rx_id) &&
+	    (rx_entry->msg_id == pkt_hdr->msg_id))
 		return 0;
 
 	return -FI_EALREADY;
 }
 
-void rxd_post_cts(struct rxd_ep *rxd_ep, struct rxd_x_entry *rx_entry,
-		  struct rxd_pkt_entry *rts_pkt_entry)
+static void rxd_transfer_pending(struct rxd_ep *ep, int peer)
 {
 	struct rxd_pkt_entry *pkt_entry;
-	struct rxd_ctrl_pkt *rts_pkt;
-	int ret;
+	struct rxd_x_entry *tx_entry;
+	struct rxd_data_pkt *data;
+	struct rxd_op_pkt *op;
+
+	while (ep->peers[peer].free_unacked &&
+	       !dlist_empty(&ep->peers[peer].pending) &&
+	       !ep->peers[peer].blocking) {
+		dlist_pop_front(&ep->peers[peer].pending,
+					struct rxd_pkt_entry, pkt_entry, d_entry); 
+		dlist_insert_tail(&pkt_entry->d_entry, &ep->peers[peer].unacked);
+		if (rxd_pkt_type(pkt_entry) == RXD_DATA) {
+			data = (struct rxd_data_pkt *) (pkt_entry->pkt);
+			tx_entry = &ep->tx_fs->entry[data->pkt_hdr.tx_id].buf;
+			data->pkt_hdr.rx_id = tx_entry->rx_id;
+		} else {
+			op = (struct rxd_op_pkt *) (pkt_entry->pkt);
+			if (op->size > rxd_ep_domain(ep)->max_inline_sz)
+				ep->peers[peer].blocking = 1;
```

Comment:
Is blocking used to indicate whether you can use inject?

Reply:
I'm available for a $50 per hour consulting fee.  Super cheap!

Reply:
You wish you were worth that much. I'll pay you in old Necco wafers and bubblegum.

Reply:
I still haven't seen any Necco wafers.